core/arch/arm/kernel/user_ta.c

```diff
@@ -770,11 +801,14 @@ static TEE_Result load_elf_from_store(const TEE_UUID *uuid,
 	if (res != TEE_SUCCESS)
 		goto out;
 
-	if (prev) {
+	if (prev)
 		elf->load_addr = prev->load_addr + prev->mobj_code->size;
-		elf->load_addr = ROUNDUP(elf->load_addr,
-					 CORE_MMU_USER_CODE_SIZE);
-	}
+	else
+		elf->load_addr = utc->stack_addr + utc->mobj_stack->size;
+
+	elf->load_addr += aslr_offset(0, CFG_TA_ASLR_MAX_OFFSET_PAGES);
```

Comment:
> Do we need to randomize the relative placement of the different ELFs?

That is something I had in mind too, but it's a bit more complicated to do so I left it for a later step.

> We should have some checks to see that we can guarantee in advance that the TA can be loaded or not. That loading sometimes fails due to using too much virtual memory isn't good.

Let me rephrase to make sure we are on the same page (no pun intended ;-)).
- Without ASLR, if a TA could be mapped once it means it can be mapped again in the future because it is always mapped in the same way
- With ASLR, each time the TA is loaded it may use a different amount of page table size so there is no guarantee a particular invocation will succeed.
But what could we do?